fl4health/strategies/fedavg_sparse_coo_tensor.py

```diff
@@ -0,0 +1,302 @@
+            results (List[Tuple[NDArrays, int]]): The weight results from each client's local training
+            that need to be aggregated on the server-side and the number of training samples
+            held on each client.
+
+                In this scheme, the clients pack the tensor names into the results object along with
+                the weight values to allow for alignment during aggregation.
+
+        Returns:
+            Dict[str, Tensor]: A dictionary mapping the name of the tensor that was aggregated to the aggregated
+                weights.
+        """
+        if self.weighted_aggregation:
+            return self.weighted_aggregate(results)
+        else:
+            return self.unweighted_aggregate(results)
+
+    def weighted_aggregate(self, results: List[Tuple[NDArrays, int]]) -> Dict[str, Tensor]:
+        """
+        "results" consist of four parts: the exchanged (nonzero) parameter values,
+        their coordinates within the tensor to which they belong,
+        the shape of that tensor, and finally the name of that tensor.
+
+        The first three items constitute the information that is needed
+        to construct the tensor in the sparse COO format and convert
+        it to a regular dense tensor.
+        The tensor name is used to align tensors to ensure that averaging is performed only
+        among tensors with the same name.
+
+        This method performs the following steps:
+            1. Align all tensors according to their names.
+            2. For tensors that have the same name, construct the sparse COO tensors and convert them to dense tensors.
+            3. Perform weighted averaging on the dense tensors according to
+            the number of training examples each client has.
+
+        Note: this method performs weighted averaging.
+
+        Args:
+            results (List[Tuple[NDArrays, int]]): The weight results from each client's local training that need to be
+                aggregated on the server-side and the number of training samples held on each client.
+                The weight results consist of four parts, as detailed above.
+                In this scheme, the clients pack the layer names into the results object along with the weight values
+                to allow for alignment during aggregation.
+
+        Returns:
+            Dict[str, Tensor]: A dictionary mapping the name of the tensor that was aggregated to the aggregated
+                weights.
+        """
+        names_to_dense_tensors: DefaultDict[str, List[Tensor]] = defaultdict(list)
+        total_num_examples: DefaultDict[str, int] = defaultdict(int)
+
+        for packed_parameters, num_examples in results:
+            nonzero_parameter_values, additional_info = self.parameter_packer.unpack_parameters(packed_parameters)
+            parameter_indices, tensor_shapes, tensor_names = additional_info
+            for tensor_params, tensor_param_indices, tensor_shape, tensor_name in zip(
```

Comment:
Since zip has burned us in the past (myself very much included), can we just do a quick assert to make sure the sizes of all of the pieces being zipped together are equal and non-zero?